### PR TITLE
Fix RPCClient Deadlock on Unsubscribe and NewHead

### PR DESCRIPTION
**Pull request by dhaidashenko**
[Ticket](https://smartcontract-it.atlassian.net/browse/BCI-4095)

Concurrent calls to `onNewHead` and `UnsubscribeAllExceptAliveLoop` caused deadlock on `stateMu`. As a result, any call to any configured RPC for that chain is blocked.

PR addresses the issue by adding a separate lock to sync operations related to ChainInfo separately from subscription management.

### CODE REVIEW

**Review by DylanTinianov — APPROVED** [review]

**Review by DylanTinianov — APPROVED** [review]
Looks good

**Review by amit-momin — APPROVED** [review]
LGTM

File: core/chains/evm/client/rpc_client_test.go [hunk 1]
```diff
@@ -130,6 +131,33 @@ func TestRPCClient_SubscribeNewHead(t *testing.T) {
 		assert.Equal(t, int64(0), highestUserObservations.FinalizedBlockNumber)
 		assert.Equal(t, (*big.Int)(nil), highestUserObservations.TotalDifficulty)
 	})
+	t.Run("Concurrent Unsubscribe and onNewHead calls do not lead to a deadlock", func(t *testing.T) {
+		const numberOfAttempts = 1000 // need a large number to increase the odds of reproducing the issue
```
Comment (reductionista) [inline core/chains/evm/client/rpc_client_test.go:135]: If this only reproduces the issue probabilistically, does that make the test flaky? Or is it just that there will be a low chance of it PASS'ing with a false positive?

**Review by prashantkumar1982 — APPROVED** [review]

**Review by huangzhen1997 — APPROVED** [review]

File: core/chains/evm/client/rpc_client_test.go (hunk 1 above)
Comment (huangzhen1997) [reply core/chains/evm/client/rpc_client_test.go:135]: probably just creating a potential race condition, and now we have the right locking, we don't expect the test to fail/deadlock happens. So it shouldn't be flaky I think

Comment (dhaidashenko) [reply core/chains/evm/client/rpc_client_test.go:135]: @reductionista, flaky behavior will only occur if there is regression. 1000 attempts seem to be large enough. Before the fix, the test failed on each run.

**Review by reductionista — APPROVED** [review]

**Pull request merged**